# Exercises Part I

In [1]:
import pandas as pd
import numpy as np
from pydataset import data
from env import host, user, password



In [21]:
def get_db_url(user,host,password):
    url = f'mysql+pymysql://{user}:{password}@{host}/employees'
    return url

In [22]:
url = get_db_url(user,host,password)

In [23]:
print(url)

mysql+pymysql://mirzakhani_1941:2pPKm8MvKyWBguaLOFtUMRwmHwqAlNZs@data.codeup.com/employees


In [27]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)


emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

## Q5

Q5a

NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:ysql.pymysql

Q5b

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FOM employees LIMIT 5 OFFSET 50' at line 1")
[SQL: SELECT * FOM employees LIMIT 5 OFFSET 50]
(Background on this error at: https://sqlalche.me/e/14/f405)

## Q6

In [32]:
employees_df = pd.read_sql('SELECT * FROM employees', url)

In [33]:
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [34]:
titles_df = pd.read_sql('SELECT * FROM titles', url)

In [35]:
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

## Q7

#### employees, 300024 rows × 6 columns
#### titles,    443308 rows × 4 columns

This is expected because the employees data base is very large and tables differ in columns as tables are often referenced to one another in the database.

## Q8

In [36]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [37]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

## Q9

In [45]:
pd.read_sql('SELECT DISTINCT title FROM titles', url)

title
0     Senior Engineer
1               Staff
2            Engineer
3        Senior Staff
4  Assistant Engineer
5    Technique Leader
6             Manager

## Q10

In [44]:
pd.read_sql('SELECT to_date FROM titles ORDER BY to_date LIMIT 10', url)

to_date
0  1985-03-01
1  1985-03-11
2  1985-03-20
3  1985-03-23
4  1985-04-17
5  1985-04-20
6  1985-04-21
7  1985-04-23
8  1985-05-05
9  1985-05-05

Oldest to_date is 1985-03-01

## Q11

In [51]:
pd.read_sql('SELECT to_date FROM titles ORDER BY to_date DESC LIMIT 10', url)

to_date
0  9999-01-01
1  9999-01-01
2  9999-01-01
3  9999-01-01
4  9999-01-01
5  9999-01-01
6  9999-01-01
7  9999-01-01
8  9999-01-01
9  9999-01-01

Most recent to_date is 9999-01-01

# Exercises Part II